10212: document send step (#10684)
CDCgov/prime-reportstream: prime-router/docs/universal-pipeline/send.md

# Send

This is the final step in the pipeline and is responsible for actually delivering the batched report to a receivers

> preferred location

Reviewer: ```suggestion
configured location
```

- SFTP send the report over SSH to a configured SFTP server, credential must be stored in the vault
  - `host`: the url for the SFTP server
  - `port`: the port the SFTP server runs on
  - `filePath`: what path on the SFTP server the report should be placed in
  - `credentialName`: the name of the credential that should be used for authenticating the request
- Soap (only works for PA) send the report to a SOAP endpoint, credential must be configured in the vault
  - `endpoint`: the URL to send the SOAP post to
  - `soapAction`: which SOAP action to invoke
  - `credentialName`: the name of the credential to find in the vault
  - `namespace`: the optional namespace in the SOAP object

Example for an SFTP configuration (most common):

```yaml
transport:
  type: SFTP
  host: sftp
  port: 22
  filePath: ./upload
  credentialName: DEFAULT-SFTP
```

## How it works

This step works by listening for messages to get added to the send queue and then looking up the transport that is configured
for the receiver.  Each transport has its own internal logic for creating the report and sending it out, but they mostly all follow

> the similar pattern of looking up the credentials for that transport/receiver, create an external file name and send the contents of the report.

Reviewer: ```suggestion
the similar pattern of looking up the credentials for that transport/receiver, creating an external file name and sending the contents of the report.
```

**Some important notes on the transports:**
- The AS2 transport is unused
- The SOAP transport is hardcoded to only work with the state of Pennsylvania
- The Email transport is incomplete and does not work

> - The Rest Transport uses the KTOR library under the hood; if it encounters a `ClientRequestException` (any 4XX status), this transport wil short circuit and

Reviewer: ```suggestion
- The Rest Transport uses the KTOR library under the hood; if it encounters a `ClientRequestException` (any 4XX status), this transport will short circuit and
```

no retries will be attempted
- The BlobStore transport only supports azure as the backing blob store and should only be used to move blobs internally in the CDC

### Retries

All the transports are configured to catch any thrown exception and return a retry token of all items.  When a retry token is returned, the SendFunction
looks to see if it should re-queue the event for another attempt at getting sent.  The SendFunction will attempt to send the report 5 more times 
using an exponential back off with a small, random amount of time added or subtracted; this is done in order to prevent a huge slew of send events all being 

> triggered at the same time in the event of a large outage.  See examples for more details.

Reviewer: `[examples](#examples)` I think will work here

### A report fails to send the first time, but is sent the second

A receiver is configured to receive reports in a SFTP server and a new message is added to the send queue for that receiver.  The send queue
parses the message and finds the relevant credentials in the vault. The function generates an external name and attempts to upload the file
to the SFTP server, but it is currently down.  The SendFunction queues a new event with a visibility timeout of 8 minutes 
(10 minutes since it is the first retry minus 2 minutes of random "dithering" time).  That second event is then read from the queue and the report is delivered.

### A report fails to send five times and is never sent

A receiver is configured to receive reports in a SFTP server and a new message is added to the send queue for that receiver.  The send queue
parses the message and finds the relevant credentials in the vault. The function generates an external name and attempts to upload the file
to the SFTP server, but gets a response that the credentials are invalid.  The function will then queue events with exponentially larger

> visibility timeouts until all five retries have occurred.  The function will then mark that the report failed to be sent.

Reviewer: Never mind this seems more appropriate in the troubleshooting doc or elsewhere, but keeping here to document my thoughts.
